### ISSUE TITLE
chain(adam, clip_by_global_norm) clips the step, so gradient spikes still land

### ISSUE THREAD

**Issue by jessegrabowski** [pymc-devs/pytensor-ml#114]
Transforms only run after the head rule, so `clip_by_global_norm` sees Adam's step rather than the gradient. Adam has already normalized that step to roughly `learning_rate` per coordinate, so the clip essentially never fires, while the spike it was meant to stop is already in `m` and `v`. There's no way to write clip-then-adam as a chain — you have to clip the gradients yourself and hand them to the rule, which is what torch, optax, pyro and numpyro all do. The docstrings for `chain` and `clip_by_global_norm` both use `chain(adam(1e-3), clip_by_global_norm(1.0))` as the worked example, so the recommended idiom is the one that doesn't work.

```python
import numpy as np
import pytensor
import pytensor.tensor as pt

from pytensor_ml.optim import adam, chain, clip_by_global_norm
from pytensor_ml.params import trainable


def run(rule, clip_gradients=None, n_steps=20, spike=1e6):
    x = trainable(np.zeros(3), name="x")
    gradient = pytensor.shared(np.ones(3), name="g")
    grads = [gradient]
    if clip_gradients is not None:
        grads = [pt.clip(g, -clip_gradients, clip_gradients) for g in grads]
    step = pytensor.function([], x, updates=rule(grads, [x]))

    gradient.set_value(np.full(3, spike))  # one exploding gradient
    step()
    gradient.set_value(np.ones(3))  # then 19 ordinary ones
    for _ in range(n_steps - 1):
        value = step()
    return value


print(run(chain(adam(1e-3), clip_by_global_norm(1.0))))  # [-0.0052 ...] stalled
print(run(adam(1e-3), clip_gradients=1.0))  # [-0.019 ...] full speed
# what you want to write, but can't: chain(clip_by_global_norm(1.0), adam(1e-3))
```

3.6x less progress over the 19 clean steps after the spike.
